keel-orca/src/main/kotlin/com/netflix/spinnaker/keel/constraints/PipelineConstraintEvaluator.kt

```diff
@@ -0,0 +1,221 @@
+import kotlinx.coroutines.runBlocking
+import org.springframework.stereotype.Component
+import java.lang.Exception
+import java.time.Clock
+import org.slf4j.LoggerFactory.getLogger
+
+/**
+ * An environment promotion constraint to gate promotions on the successful execution
+ * of a Spinnaker pipeline. The target pipeline is launched via orca the first time
+ * a pipeline constraint is evaluated for a given artifact version and [Environment].
+ *
+ * The state of that pipeline execution is then assessed on each subsequent evaluation
+ * and can optionally be retried on failure.
+ *
+ * Example env constraint:
+ *
+ * constraints:
+ *  - type: pipeline
+ *    serviceAccount: keel@spinnaker (OPTIONAL, defaults to keel's `DEFAULT_SERVICE_ACCOUNT`)
+ *    pipelineId: (the id visible when editing a pipeline, not its name)
+ *    retries: (OPTIONAL, the number of times to retry a failure, defaults to 0)
+ *    parameters: (OPTIONAL, Map of trigger parameters to pass along, `user` and `type` are reserved)
+ *
+ */
+@Component
+class PipelineConstraintEvaluator(
+  private val orcaService: OrcaService,
+  override val deliveryConfigRepository: DeliveryConfigRepository,
+  private val clock: Clock
+) : StatefulConstraintEvaluator<PipelineConstraint>() {
+  override val constraintType = PipelineConstraint::class.java
+  private val log by lazy { getLogger(javaClass) }
+
+  override fun canPromote(
+    artifact: DeliveryArtifact,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    targetEnvironment: Environment,
+    constraint: PipelineConstraint,
+    state: ConstraintState
+  ): Boolean {
+    var attributes = state.attributes as PipelineConstraintStateAttributes?
+    val status = pipelineStatus(attributes)
+
+    if (attributes != null && status != null && attributes.lastExecutionStatus != status.toString()) {
+      attributes = attributes.copy(lastExecutionStatus = status.toString())
+    }
+
+    val judge = "pipeline:${constraint.pipelineId}"
+
+    // TODO: if the constraint has timed out but the pipeline is still running, should we cancel it?
+    if (timedOut(constraint, state, attributes)) {
+      deliveryConfigRepository
+        .storeConstraintState(
+          state.copy(
+            status = FAIL,
+            attributes = attributes,
+            judgedBy = judge,
+            judgedAt = clock.instant(),
+            comment = "Timed out after ${constraint.timeout} and ${attributes?.attempt ?: "0"} attempt" +
+              if (attributes?.attempt != 1) {
+                "s"
+              } else {
+                ""
+              }))
+      // TODO: Emit event
+      return false
+    }
+
+    if (shouldTrigger(constraint, attributes)) {
+      try {
+        val executionId = runBlocking {
+          startPipeline(targetEnvironment, constraint)
+        }
+
+        attributes = PipelineConstraintStateAttributes(
+          executionId = executionId,
+          attempt = attributes?.attempt ?: 0 + 1,
```

Comment:
Is `0 + 1` doing something other than returning 1?

Reply:
I interpreted this as `(attributes?.attempt ?: 0) + 1`, but maybe not?

Reply:
This is supposed to mean what @luispollo interpreted it as but turns out it actually does need the parenthesis, will fix.